[Property Editor] Emphasize properties the user has set in their code

elliette opened the pull request:
Fixes https://github.com/flutter/devtools/issues/8653
Work towards https://github.com/flutter/devtools/issues/1948

We would like to highlight the properties that the user has set in their code so that its visually clear what they have set.

I swapped the `required` label with a `set` label, and moved required to the input indicator (following the [Material 3 guidance](https://m3.material.io/components/text-fields/guidelines#6c36a812-273d-4f5d-9102-2f00386c9145)):

<img width="406" alt="Screenshot 2025-01-14 at 2 34 17 PM" src="https://github.com/user-attachments/assets/2e49982f-357d-43e4-9fab-f029bc762463" />



**Screenshots of various themes (NOTE: before `required` label was changed):**
<img width="486" alt="Screenshot 2024-12-17 at 5 28 14 PM" src="https://github.com/user-attachments/assets/6cc46c02-f602-40c1-9908-59c212da34f5" />
<img width="488" alt="Screenshot 2024-12-17 at 5 27 56 PM" src="https://github.com/user-attachments/assets/42c463e0-2dfd-4429-9c3d-e0ce1126b0b2" />
<img width="487" alt="Screenshot 2024-12-17 at 5 27 34 PM" src="https://github.com/user-attachments/assets/29f78fbc-5bae-4407-99bb-92c126dbc11b" />
<img width="484" alt="Screenshot 2024-12-17 at 5 27 15 PM" src="https://github.com/user-attachments/assets/e71e171e-9df2-421a-9dfa-007d06ee61df" />

<img width="481" alt="Screenshot 2024-12-17 at 5 26 58 PM" src="https://github.com/user-attachments/assets/53505a2e-3679-4760-aef6-92a71aa9aca5" />

kenzieschmoll commented:
This makes the non-set properties look disabled IMO. Is this something we could put on our list to get UX input on? what about another badge "explicit" or "set" instead of graying out some of the properties?

elliette commented:
> This makes the non-set properties look disabled IMO. Is this something we could put on our list to get UX input on? what about another badge "explicit" or "set" instead of graying out some of the properties?

We already have 2 potential badges (isDefault and isRequired), and I think we will need to change the UI to fit 3. 

I've tweaked how we are determining the background color, and added a bunch of different screenshots for various VS Code themes. Let me know what you think - mostly I want this in a nice-ish state for the demo, but I agree this is something we should follow up on with UX.

Note: I haven't added comments and tests yet, will add those if we think we should go ahead with these colors for now.

kenzieschmoll commented:
What about something more subtle like this?
![Screenshot 2024-12-17 at 3 23 41 PM](https://github.com/user-attachments/assets/7ecfd6db-8009-4bdf-92ac-30dd1c1a46a9)

kenzieschmoll commented:
Another affordance to consider if we want to reduce the number of badges like "default". We could use an asterisk `*` to indicate required parameters. Kind of like you'd expect to see in a form for required fields. Then maybe that would open up more space for a "set" or "explicit" badge if we think that would be more clear than using color.

elliette commented:
> What about something more subtle like this? ![Screenshot 2024-12-17 at 3 23 41 PM](https://private-user-images.githubusercontent.com/43759233/396697896-7ecfd6db-8009-4bdf-92ac-30dd1c1a46a9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQ0ODIxOTcsIm5iZiI6MTczNDQ4MTg5NywicGF0aCI6Ii80Mzc1OTIzMy8zOTY2OTc4OTYtN2VjZmQ2ZGItODAwOS00YmRmLTkyYWMtMzBkZDFjMWE0NmE5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjE4VDAwMzEzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM2YmUyNTViMTA2MGFjMWIxNTI3YTIxZjI5NDFmM2JiY2U3MDI4YWIwODg1MGUzNWFhMzRhNmY3MDY3ODczYWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.8XlcIbScs7KbUr4h-fzLYPC-YWmmBgLAkFcJdSDJz64)

Oh I like this option! Changed to that and updated the screenshots.

kenzieschmoll commented on packages/devtools_app/lib/src/standalone_ui/ide_shared/property_editor/property_editor_view.dart:
```diff
@@ -7,7 +7,6 @@ import 'package:flutter/material.dart';
-import '../../../shared/primitives/utils.dart';
 import 'property_editor_controller.dart';
 
 class PropertyEditorView extends StatelessWidget {
```
this widget doesn't provide much value to abstract now. Can we add the `_PropertiesList` directly where `PropertyEditorView` is currently used? That or make `PropertyEditorView` what `_PropertiesList` currently is (effectively deleting `_PropertiesList` and having `PropertyEditorView` contain the body of the current `_PropertiesList`)

elliette replied:
Yes, we can! Though I'm planning on adding more to the `PropertyEditorView` soon (for example, the widget name) so will probably end up adding it back in later. Do you still think I should delete it for now?

elliette replied:
I can add a TODO for the parts that are coming soon

kenzieschmoll commented on packages/devtools_app/lib/src/standalone_ui/ide_shared/property_editor/property_editor_view.dart:
```diff
@@ -72,23 +70,48 @@ class _EditablePropertyItem extends StatelessWidget {
+        ),
+        child: Row(
+          crossAxisAlignment: CrossAxisAlignment.start,
+          children: [
```
nit: it might be simpler to just add a Container to this row directly:
```
Row(
  children: [
    Container(
      width: _hasArgIndicatorWidth,
      color: argument.hasArgument ? theme.colorScheme.primary : Colors.transparent,
    ),
    ...
  ],
),
```

elliette replied:
Hmm I tried that and was struggling to get it the proper height which is why I switched to border instead

kenzieschmoll commented on packages/devtools_app/lib/src/standalone_ui/ide_shared/property_editor/property_editor_view.dart:
```diff
@@ -143,11 +146,12 @@ class _PropertyInputState extends State<_PropertyInput> {
+      errorText: argument.errorText,
       isDense: true,
-      label: Text(widget.argument.name),
+      label: Text('${argument.name}${argument.isRequired ? '* ' : ''}'),
```
do we want a trailing space after the * char?

elliette replied:
good point, no we don't!
